### PR TITLE
gh-102978: Fix mock.patch function signatures for class and staticmethod decorators

### LINKED ISSUE #102978: unittest.mock.patch with autospec doesn't hold for classmethods nor staticmethods
# Bug report

Using `patch` with `autospec=True` does not work when a method is decorated with `@classmethod` or `@staticmethod`. The resulting mock can be called with any arguments without raising a `TypeError`.

Example:
```python
from unittest.mock import patch
import pytest

class Foo:
    def foo(self):
        pass
    @staticmethod
    def bar():
        pass
    @classmethod
    def baz(cls):
        pass

@pytest.mark.parametrize("method", ["foo", "bar", "baz"])
def test_foo(method: str):
    with patch.object(Foo, method, autospec=True):
        getattr(Foo(), method)(5)
```

The only subtest that fails is `foo`. The other two pass, even though they're clearly being called incorrectly.

If you prefer not to use `pytest` to demo/repro this:

```python
with patch.object(Foo, "foo", autospec=True):
    try:
        Foo().foo(5)
    except TypeError:
        print("Correctly raises on foo")
    else:
        print("Incorrectly does not raise with foo")

with patch.object(Foo, "bar", autospec=True):
    try:
        Foo().bar(5)
    except TypeError:
        print("Correctly raises on bar")
    else:
        print("Incorrectly does not raise with bar")

with patch.object(Foo, "baz", autospec=True):
    try:
        Foo().baz(5)
    except TypeError:
        print("Correctly raises on baz")
    else:
        print("Incorrectly does not raise with baz")
```

This has output:
```
Correctly raises on foo
Incorrectly does not raise with bar
Incorrectly does not raise with baz
```

# Your environment

- CPython versions tested on: 3.10, 3.11
- Operating system and architecture: macOS 12.6 with Apple M1 chip

### Linked PRs
* gh-103228
* gh-103499

### PR DESCRIPTION
**Pull request by tomasr8**
Closes #102978 

As reported in the original [issue](https://github.com/python/cpython/issues/102978), `unittest.mock.patch` does not correctly extract function signatures from methods decorated with `@classmethod` and `@staticmethod` when `patch` is called with `autospec=True`.

This fix adds a special case to the extraction logic which uses the original decorated function via `__func__` to get the correct signature.

I also updated the relevant tests.

* Issue: gh-102978

### CODE REVIEW

File: Lib/unittest/mock.py [hunk 1]
```diff
@@ -98,6 +98,12 @@ def _get_signature_object(func, as_instance, eat_self):
         func = func.__init__
         # Skip the `self` argument in __init__
         eat_self = True
+    elif isinstance(func, classmethod | staticmethod):
```
Comment (sobolevn) [inline Lib/unittest/mock.py]:
```suggestion
    elif isinstance(func, (classmethod,  staticmethod)):
```

We should stick to tuples in this case :)

File: Misc/NEWS.d/next/Library/2023-04-03-23-44-34.gh-issue-102978.gy9eVk.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Fixes unittest.mock.patch not enforcing function signatures for methods
+decorated with @classmethod or @staticmethod when patch is called with
+autospec=True.
```
Comment (sobolevn) [inline Misc/NEWS.d/next/Library/2023-04-03-23-44-34.gh-issue-102978.gy9eVk.rst]:
```suggestion
Fixes :func:`unittest.mock.patch` not enforcing function signatures for methods
decorated with ``@classmethod`` or ``@staticmethod`` when patch is called with
``autospec=True``.
```

File: Lib/test/test_unittest/testmock/testpatch.py [hunk 3]
```diff
@@ -989,12 +989,18 @@ def test_autospec_staticmethod(self):
...
+            self.assertRaises(TypeError, Foo.static_method, 42)
+
 
     def test_autospec_classmethod(self):
         with patch('%s.Foo.class_method' % __name__, autospec=True) as method:
             Foo.class_method()
             method.assert_called_once_with()
 
+            # Ensure that the original method's signature is checked correctly
+            self.assertRaises(TypeError, Foo.class_method, 42)
```
Comment (sobolevn) [inline Lib/test/test_unittest/testmock/testpatch.py]: I think that this test is not enough. Please, create a new test with class and static methods that have some parameters. And check that these parameters do work :)

**Review by sobolevn — COMMENTED** [review]
Thanks!

File: Lib/test/test_unittest/testmock/testpatch.py (hunk 3 above)
Comment (tomasr8) [reply Lib/test/test_unittest/testmock/testpatch.py]: Thanks for reviewing this PR! I added more tests :)

File: Lib/test/test_unittest/testmock/testhelpers.py [hunk 4]
```diff
@@ -952,6 +952,30 @@ def __getattr__(self, attribute):
         self.assertFalse(hasattr(autospec, '__name__'))
 
 
+    def test_autospec_signature_staticmethod(self):
+        # Regression test for gh-102978
+        class Foo:
+            @staticmethod
+            def static_method(a, b=10, *, c): pass
+
+        Foo.static_method(1, 2, c=3)
```
Comment (sobolevn) [inline Lib/test/test_unittest/testmock/testhelpers.py]: We don't really need this line

File: Lib/test/test_unittest/testmock/testhelpers.py [hunk 5]
```diff
@@ -952,6 +952,30 @@ def __getattr__(self, attribute):
...
+        self.assertEqual(inspect.signature(Foo.static_method), inspect.signature(mock))
+
+
+    def test_autospec_signature_classmethod(self):
+        # Regression test for gh-102978
+        class Foo:
+            @classmethod
+            def class_method(cls, a, b=10, *, c): pass
+
+        Foo.class_method(1, 2, c=3)
```
Comment (sobolevn) [inline Lib/test/test_unittest/testmock/testhelpers.py]: And this

**Review by sobolevn — COMMENTED** [review]
Almost done, in my opinion :)

**Comment by tomasr8** [comment]
Thanks! I updated the tests

**Review by sobolevn — COMMENTED** [review]
LGTM, but you will need an approve from the code owner :)

**Comment by tomasr8** [comment]
Awesome, thanks for your time!

File: Lib/test/test_unittest/testmock/testhelpers.py [hunk 6]
```diff
@@ -952,6 +952,26 @@ def __getattr__(self, attribute):
         self.assertFalse(hasattr(autospec, '__name__'))
 
 
+    def test_autospec_signature_staticmethod(self):
+        # Regression test for gh-102978
```
Comment (cjw296) [inline Lib/test/test_unittest/testmock/testhelpers.py]: It's a small thing, but these comments don't add much and will bitrot.
```suggestion
```

File: Lib/test/test_unittest/testmock/testhelpers.py [hunk 7]
```diff
@@ -952,6 +952,26 @@ def __getattr__(self, attribute):
...
+        class Foo:
+            @staticmethod
+            def static_method(a, b=10, *, c): pass
+
+        mock = create_autospec(Foo.__dict__['static_method'])
+        self.assertEqual(inspect.signature(Foo.static_method), inspect.signature(mock))
+
+
+    def test_autospec_signature_classmethod(self):
+        # Regression test for gh-102978
```
Comment (cjw296) [inline Lib/test/test_unittest/testmock/testhelpers.py]:
```suggestion
```

File: Lib/test/test_unittest/testmock/testpatch.py [hunk 8]
```diff
@@ -996,6 +996,38 @@ def test_autospec_classmethod(self):
             method.assert_called_once_with()
 
 
+    def test_autospec_staticmethod_signature(self):
+        # Regression test for gh-102978
+        # Patched methods which are decorated with @staticmethod should have the same signature
```
Comment (cjw296) [inline Lib/test/test_unittest/testmock/testpatch.py]:
```suggestion
        # Patched methods which are decorated with @staticmethod should have the same signature
```

File: Lib/test/test_unittest/testmock/testpatch.py [hunk 9]
```diff
@@ -996,6 +996,38 @@ def test_autospec_classmethod(self):
...
+        with patch.object(Foo, 'static_method', autospec=True) as method:
+            method(1, 2, c=3)
+            self.assertRaises(TypeError, method)
+            self.assertRaises(TypeError, method, 1)
+            self.assertRaises(TypeError, method, 1, 2, 3, c=4)
+
+
+    def test_autospec_classmethod_signature(self):
+        # Regression test for gh-102978
+        # Patched methods which are decorated with @classmethod should have the same signature
```
Comment (cjw296) [inline Lib/test/test_unittest/testmock/testpatch.py]:
```suggestion
        # Patched methods which are decorated with @classmethod should have the same signature
```

**Review by cjw296 — APPROVED** [review]
Really nice PR, thanks!

**Pull request merged**